### PR TITLE
chore(raft): introduce snapshot metrics and utilities

### PR DESCRIPTION
**Pull request by npepinpe**
- adds snapshot replication metrics
- adds SnapshotListener#onSnapshotDeletion (allows tracking removal)

### CODE REVIEW

**Comment by npepinpe** [comment]
Here is how it now looks in Grafana:

Screenshot from 2020-01-15 14-53-21

File: protocols/raft/src/main/java/io/atomix/protocols/raft/metrics/SnapshotReplicationMetrics.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          .labelNames(PARTITION_GROUP_NAME_LABEL, PARTITION_LABEL)
+          .help("Count of ongoing snapshot replication")
+          .name("pending_snapshot_count")
+          .register();
+  private static final Gauge DURATION =
+      Gauge.build()
+          .namespace(NAMESPACE)
+          .labelNames(PARTITION_GROUP_NAME_LABEL, PARTITION_LABEL)
+          .help("Approximate duration of replication in milliseconds")
+          .name("pending_snapshot_milliseconds")
```
Comment (deepthidevaki) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/metrics/SnapshotReplicationMetrics.java]: The name `pending_snapshot_milliseconds` is confusing. Better something like `snapshot_replication_duration_milliseconds`.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
:+1: just one comment.

**Review by deepthidevaki — APPROVED** [review]

**Comment by npepinpe** [comment]
bors retry

**Pull request merged**